HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaProducer.cc

```diff
@@ -27,29 +29,23 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
     TestAlpakaProducer(edm::ParameterSet const& config)
         : deviceToken_{produces()}, size_{config.getParameter<int32_t>("size")} {}
 
-    void beginStream(edm::StreamID sid) override {
-      // choose a device based on the EDM stream number
+    void beginStream(edm::StreamID) override {
```

Comment:
This function is not really needed anymore.

Reply:
OK, this is one of the open questions I had: do we need to declare some kind of dependency on the `AlpakaService`s ? do we need to do anything if the service is not available ?

Reply:
Good question. In the CUDA model the `chooseDevice()` asks the number of devices from `CUDAService`, and throws an exception if the `CUDAService` is enabled.
https://github.com/cms-sw/cmssw/blob/4c7b44b77f2ae84e06b6eb717f5822e1f19b6461/HeterogeneousCore/CUDACore/src/chooseDevice.cc#L8-L25

In principle we could do the same here, or have the `ScopedContext` classes to check that the relevant `AlpakaService` is enabled. But both of those are templated on the Queue type, whereas `AlpakaService`s are differentiated by `ALPAKA_ACCELERATOR_NAMESPACE`.

Maybe better to keep the `beginStream()` with the check for now and think this question some more?

Reply:
A possibility would be to template  the `AlpakaService` on the accelerator type (or maybe the device type) instead of putting it in the `ALPAKA_ACCELERATOR_NAMESPACE`.

Reply:
I've thought a little bit about the relation of modules and `AlpakaService`s, and I'd say that if a module does not explicitly need something from the Service, it should not depend on the Service. Presently, given that the caching allocators and queue and event caches are true singletons, an Alpaka module would work fine without the presence of the corresponding `AlpakaService` (in principle the CUDA-backend module would work also without `CUDAService`). If (when) the Services are needed in the job for other reasons, something else should ensure the corresponding Service existence in the job. (of course, if in the future any of the caches would become owned by the `AlpakaService`, then the modules would get a direct dependence on the Service)

I think one reason why the `CUDAService::numberOfDevices()` was used in the `cms::cuda::chooseDevice()` was to have a possibility to disable CUDA use in the configuration (before `process.options.accelerators`) and this use adds a cross check on that. We can (well, will) have the `process.options.accelerators` to drive any automated decisions (like `SwitchProducer`) on what backend to run. That leaves a question on what would it mean if `process.options.accelerators` tells to not use a GPU, but the process has a module that explicitly uses the CUDA backend. Maybe it should be an error condition (such configuration would anyway fail to run on a machine without a GPU). Using the corresponding `AlpakaService` in the `produce()`/`acquire()` function (hidden within the accelerator framework) would do that job, but we could think of some other mechanism as well.

Anyway, for now (to get started) the exact dependence model probably doesn't matter much, and we'll evolve it in some direction later. We just have to remember that the `beginStream()` is called for all modules (that are either explicitly scheduled, i.e. in any Path, or unscheduled and any data dependence chain from a scheduled module leads to the unscheduled module).

> A possibility would be to template the `AlpakaService` on the accelerator type (or maybe the device type) instead of putting it in the `ALPAKA_ACCELERATOR_NAMESPACE`.

Yeah. After playing around I've came to appreciate the simplicity of having code in `ALPAKA_ACCELERATOR_NAMESPACE` for code where it is sufficient to have exactly one "template instantiation" per backend. So I'd prefer to keep the `AlpakaService`s in the namespace for now.

Reply:
> After playing around I've came to appreciate the simplicity of having code in `ALPAKA_ACCELERATOR_NAMESPACE` for code where it is sufficient to have exactly one "template instantiation" per backend. So I'd prefer to keep the `AlpakaService`s in the namespace for now.

OK.

Another possibility would be something like
```c++
namespace traits {
  template <typename TAcc, typename TQueue>
  class AlpakaServiceTrait {};

#if defined ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
  template <typename TDim>
  class AlpakaServiceTrait<AccCpuSerial<TDim, Idx>, QueueCpuBlocking> {
    using type = alpaka_serial_sync::AlpakaService;
  };
#endif

#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
  template <typename TDim>
  class AlpakaServiceTrait<AccGpuCudaRt<TDim, Idx>, QueueCudaRtNonBlocking> {
    using type = alpaka_cuda_async::AlpakaService;
  };
#endif

// etc.
}

template <typename TAcc, typename TQueue>
using AlpakaService = traits::AlpakaServiceTrait<TAcc, TQueue>::type;
```

so code outside of `ALPAKA_ACCELERATOR_NAMESPACE` could use `AlpakaService<Acc1D, Queue>`.